src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,426 @@
+        {
+            BeforeExecuteCommand(args, executionContext);
+            nextCommandHandler();
+        }
+
+        private void BeforeExecuteCommand(TypeCharCommandArgs args, CommandExecutionContext executionContext)
+        {
+            if (args.TypedChar != ';')
+            {
+                return;
+            }
+
+            var caretOpt = args.TextView.GetCaretPoint(args.SubjectBuffer);
+            if (!caretOpt.HasValue)
+            {
+                return;
+            }
+
+            var caret = caretOpt.Value;
+
+            var document = caret.Snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+
+            {
+                return;
+            }
+
+            var root = document.GetSyntaxRootSynchronously(executionContext.OperationContext.UserCancellationToken);
+            var caretPosition = caret.Position;
+
+            var token = root.FindToken(caretPosition);
+
+            var currentNode = token.Parent;
+
+            // If cursor is right before an opening delimiter, start with node outside of delimiters since analysis 
+            // starting with a node containing delimiters assumes the caret is placed inside those delimiters. 
+            // This covers cases like `obj.ToString$()`, where `token` references `(` but the caret isn't actually 
+            // inside the argument list.
+            if (token.IsKind(SyntaxKind.OpenBraceToken, SyntaxKind.OpenBracketToken, SyntaxKind.OpenParenToken)
+                && token.Span.Start >= caretPosition)
+            {
+                currentNode = currentNode.Parent;
+            }
+
+            if (currentNode == null)
+            {
+                return;
+            }
+
+            var isCaretAtEndOfLine = ((SnapshotPoint)caret).Position == ((SnapshotPoint)caret).GetContainingLine().End;
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            if (!LooksLikeNodeInArgumentList(currentNode, caret, isCaretAtEndOfLine, syntaxFacts))
+            {
+                return;
+            }
+
+            // verify all delimeters exist until you reach statement syntax that requires a semicolon
+            while (!IsStatementOrFieldDeclaration(currentNode, syntaxFacts))
+            {
+                if (!ClosingDelimiterExistsIfNeeded(currentNode))
+                {
+                    // A required delimiter is missing; do not treat semicolon as statement completion
+                    return;
+                }
+
+                if (currentNode.Parent == null)
+                {
+                    return;
+                }
+
+                currentNode = currentNode.Parent;
+            }
+
+            // if the statement syntax itself requires a closing delimeter, verify it is there
+            if (StatementClosingDelimiterIsMissing(currentNode))
+            {
+                // Example: missing final `)` in `do { } while (x$$`
```

Comment:
i don't really understand the feature here.  Isn't the purpose to insert missing delimiters?  So why don't we just insert the missing delimiters here?  I'm not saying there isn't a good reason to have the behavior you have here.  What i am saying is that hte comment doesn't *explain* the situation enough for me to understand why this is the *preferred* behavior.

Reply:
➡️ Covered by https://github.com/dotnet/roslyn/pull/30743#discussion_r242937849